--- a/arch/mips/ath79/Kconfig
+++ b/arch/mips/ath79/Kconfig
@@ -283,6 +283,19 @@ config ATH79_MACH_WPE72
 	select ATH79_DEV_USB
 	select MYLOADER
 
+config ATH79_MACH_WPJ344
+        bool "Compex WPJ344 board support"
+        select SOC_AR934X
+        select ATH79_DEV_ETH
+        select ATH79_DEV_GPIO_BUTTONS
+        select ATH79_DEV_LEDS_GPIO
+        select ATH79_DEV_M25P80
+        select ATH79_DEV_USB
+        select ATH79_DEV_WMAC
+	help
+	  Say 'Y' here if you want your kernel to support the
+	  Atheros WPJ344 reference board.
+
 config ATH79_MACH_DIR_505_A1
 	bool "D-Link DIR-505-A1 support"
 	select SOC_AR933X
--- a/arch/mips/ath79/Makefile
+++ b/arch/mips/ath79/Makefile
@@ -133,6 +133,7 @@ obj-$(CONFIG_ATH79_MACH_WNR2000_V3)	+= m
 obj-$(CONFIG_ATH79_MACH_WNR2200)	+= mach-wnr2200.o
 obj-$(CONFIG_ATH79_MACH_WP543)		+= mach-wp543.o
 obj-$(CONFIG_ATH79_MACH_WPE72)		+= mach-wpe72.o
+obj-$(CONFIG_ATH79_MACH_WPJ344)		+= mach-wpj344.o
 obj-$(CONFIG_ATH79_MACH_WRT160NL)	+= mach-wrt160nl.o
 obj-$(CONFIG_ATH79_MACH_WRT400N)	+= mach-wrt400n.o
 obj-$(CONFIG_ATH79_MACH_WZR_HP_G300NH)	+= mach-wzr-hp-g300nh.o
--- a/arch/mips/ath79/machtypes.h
+++ b/arch/mips/ath79/machtypes.h
@@ -175,6 +175,7 @@ enum ath79_mach_type {
 	ATH79_MACH_WNR612_V2,		/* NETGEAR WNR612 v2 */
 	ATH79_MACH_WP543,		/* Compex WP543 */
 	ATH79_MACH_WPE72,		/* Compex WPE72 */
+	ATH79_MACH_WPJ344,		/* Compex WPJ344 */
 	ATH79_MACH_WRT160NL,		/* Linksys WRT160NL */
 	ATH79_MACH_WRT400N,		/* Linksys WRT400N */
 	ATH79_MACH_WZR_HP_AG300H,	/* Buffalo WZR-HP-AG300H */
--- a/include/linux/ar8216_platform.h
+++ b/include/linux/ar8216_platform.h
@@ -120,6 +120,7 @@ struct ar8327_platform_data {
 	struct ar8327_sgmii_cfg *sgmii_cfg;
 	struct ar8327_port_cfg port0_cfg;
 	struct ar8327_port_cfg port6_cfg;
+        struct ar8327_port_cfg cpuport_cfg;
 	struct ar8327_led_cfg *led_cfg;
 
 	int (*get_port_link)(unsigned port);
@@ -128,4 +129,4 @@ struct ar8327_platform_data {
 	const struct ar8327_led_info *leds;
 };
 
-#endif /* AR8216_PLATFORM_H */
\ No newline at end of file
+#endif /* AR8216_PLATFORM_H */
