--- a/drivers/net/phy/ar8216.c
+++ b/drivers/net/phy/ar8216.c
@@ -1454,8 +1454,8 @@ ar8327_hw_config_pdata(struct ar8xxx_pri
 	data->port6_status = ar8327_get_port_init_status(&pdata->port6_cfg);
 
 	t = ar8327_get_pad_cfg(pdata->pad0_cfg);
-	if (chip_is_ar8337(priv))
-		t |= AR8337_PAD_MAC06_EXCHANGE_EN;
+	//if (chip_is_ar8337(priv))
+	//	t |= AR8337_PAD_MAC06_EXCHANGE_EN;
 
 	priv->write(priv, AR8327_REG_PAD0_MODE, t);
 	t = ar8327_get_pad_cfg(pdata->pad5_cfg);
