--- a/arch/mips/ath79/Kconfig
+++ b/arch/mips/ath79/Kconfig
@@ -323,6 +323,20 @@ config ATH79_MACH_WPJ558
           Say 'Y' here if you want your kernel to support the
           Compex WPJ558 board.
 
+config ATH79_MACH_WPJ531
+        bool "Compex WPJ531 board support"
+        select SOC_QCA953X
+        select ATH79_DEV_ETH
+        select ATH79_DEV_GPIO_BUTTONS
+        select ATH79_DEV_LEDS_GPIO
+        select ATH79_DEV_M25P80
+        select ATH79_DEV_USB
+        select ATH79_DEV_WMAC
+        help
+          Say 'Y' here if you want your kernel to support the
+          Compex WPJ531 board.
+
+
 config ATH79_MACH_DIR_505_A1
 	bool "D-Link DIR-505-A1 support"
 	select SOC_AR933X
--- a/arch/mips/ath79/Makefile
+++ b/arch/mips/ath79/Makefile
@@ -136,6 +136,7 @@ obj-$(CONFIG_ATH79_MACH_WPE72)		+= mach-
 obj-$(CONFIG_ATH79_MACH_WPJ342)		+= mach-wpj342.o
 obj-$(CONFIG_ATH79_MACH_WPJ344)		+= mach-wpj344.o
 obj-$(CONFIG_ATH79_MACH_WPJ558)         += mach-wpj558.o
+obj-$(CONFIG_ATH79_MACH_WPJ531)		+= mach-wpj531.o
 obj-$(CONFIG_ATH79_MACH_WRT160NL)	+= mach-wrt160nl.o
 obj-$(CONFIG_ATH79_MACH_WRT400N)	+= mach-wrt400n.o
 obj-$(CONFIG_ATH79_MACH_WZR_HP_G300NH)	+= mach-wzr-hp-g300nh.o
--- a/arch/mips/ath79/machtypes.h
+++ b/arch/mips/ath79/machtypes.h
@@ -178,6 +178,7 @@ enum ath79_mach_type {
 	ATH79_MACH_WPJ342,		/* Compex WPJ342 */
 	ATH79_MACH_WPJ344,		/* Compex WPJ344 */
         ATH79_MACH_WPJ558,              /* Compex WPJ558 */
+        ATH79_MACH_WPJ531,              /* Compex WPJ531 */
 	ATH79_MACH_WRT160NL,		/* Linksys WRT160NL */
 	ATH79_MACH_WRT400N,		/* Linksys WRT400N */
 	ATH79_MACH_WZR_HP_AG300H,	/* Buffalo WZR-HP-AG300H */
